Diagnostics say that the `let` in let chains has type bool

### Code

```Rust
fn main() {
    if let x = 1 && 2 {}
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:2:21
  |
2 |     if let x = 1 && 2 {}
  |        ---------    ^ expected `bool`, found integer
  |        |
  |        expected because this is `bool`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (bin "playground") due to 1 previous error
```

### Desired output

```Shell

```

### Rationale and extra context

The compiler is saying that `let x = 1` has type `bool`, which doesn't make much sense.

### Other cases

```Rust

```

### Rust Version

```Shell
Reproducible on the playground with version: 1.92.0-nightly (2025-10-13 4b94758d2ba7d0ef71cc)
```

### Anything else?

Discovered by ashtree129 on the community discord.

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_typeck/src/demand.rs
+++ b/compiler/rustc_hir_typeck/src/demand.rs
@@ -792,7 +792,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 hir::Node::Expr(hir::Expr { kind: hir::ExprKind::Binary(_, lhs, rhs), .. }),
                 Some(TypeError::Sorts(ExpectedFound { expected, .. })),
             ) if rhs.hir_id == expr.hir_id
-                && self.typeck_results.borrow().expr_ty_adjusted_opt(lhs) == Some(expected) =>
+                && self.typeck_results.borrow().expr_ty_adjusted_opt(lhs) == Some(expected)
+                // let expressions being marked as `bool` is confusing (see issue #147665)
+                && !matches!(lhs.kind, hir::ExprKind::Let(..)) =>
             {
                 err.span_label(lhs.span, format!("expected because this is `{expected}`"));
             }

--- a/tests/ui/binop/let-chain-type-issue-147665.rs
+++ b/tests/ui/binop/let-chain-type-issue-147665.rs
@@ -1,0 +1,7 @@
+// Shouldn't highlight `let x = 1` as having type bool.
+//@ edition:2024
+
+fn main() {
+    if let x = 1 && 2 {}
+    //~^ ERROR mismatched types
+}

--- a/tests/ui/binop/let-chain-type-issue-147665.stderr
+++ b/tests/ui/binop/let-chain-type-issue-147665.stderr
@@ -1,0 +1,9 @@
+error[E0308]: mismatched types
+  --> $DIR/let-chain-type-issue-147665.rs:5:21
+   |
+LL |     if let x = 1 && 2 {}
+   |                     ^ expected `bool`, found integer
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

